dash.el

```diff
@@ -2136,6 +2136,66 @@ because we need to support improper list binding."
       `(let ,inputs
          (-let* ,new-varlist ,@body)))))
 
+(defun dash--make-arglist (args)
+  "Make ARGS a function arglist for `dash--destructure-arglist'."
+  (--map-indexed
+   ;; No need to destructure a symbol to itself.
+   (if (symbolp it) it (intern (format "input%d" it-index))) args))
+
+(defun dash--progn (body-forms)
+  "Wrap BODY-FORMS in a `progn'.
+The `progn' is omitted if it isn't needed. This function is just
+`macroexp-progn' reimplemented."
+  (if (cdr body-forms)
+      `(progn ,@body-forms)
+    (car body-forms)))
+
+(defun dash--destructure-arglist (args body-forms)
+  "Destructure function arguments ARGS using `-let'.
+The result, an ELisp-form, is valid only in a function whose
+arguments are the result of transforming ARGS with
+`dash--make-arglist'. In the resulting form, BODY-FORMS are
+executed within the bound environment."
+  (unless (listp args)
+    (signal 'wrong-type-argument "match-form must be a list"))
+  (let ((let-bindings
+         (-remove #'null
+                  (--map-indexed
+                   ;; Symbols shouldn't be rebound; they can be taken from the
+                   ;; surrounding environment directly.
+                   (unless (symbolp it)
+                     (list it (intern (format "input%d" it-index))))
+                   args))))
+    (if let-bindings
+        `(-let ,let-bindings ,@body-forms)
```

Comment:
Note the `-let`: previously, `-let*` was used instead, which might be more efficient as the latter is implemented in terms of the latter. Not doing that might also make my various symbol and &as optimizations for naught. What do you think?

Reply:
I don't think I can say something intelligent on this, as I have not yet looked at the code closely.

But re: `-let` being implemented in terms of `-let*` - I think that's an implementation detail that we should not base other design decisions on, since `-let` would be implemented only in terms of itself.

Reply:
You need to make sure that input arguments are evaluated in parallel, whether with `-let` or `-let*` doesn't matter much.

Reply:
You would need to construct some kind of graph of dependencies between the bindings.  Which I suppose was too much work and I simply leveraged the built-in let to handle it for me.  But yes this is a possible optimization.  In practice probably insignificant :)  But it could be intelectually stimulating :D

You're right about the lambda, there the parallel eval is handled by the runtime when the arguments are passed to the "wrapper lambda".